core/src/main/scala/chisel3/BlackBox.scala

```diff
@@ -133,23 +133,27 @@ package experimental {
   * @note The parameters API is experimental and may change
   */
 abstract class BlackBox(val params: Map[String, Param] = Map.empty[String, Param])(implicit compileOptions: CompileOptions) extends BaseBlackBox {
+  @deprecated("For type-safe interfaces, provide your own trait. For reflective API, use DataMirror", "Chisel 3.4")
```

Comment:
Consider adding a bit more description here, something to the effect of "Module.io will be removed since it breaks in Scala 2.12+; you can still define io Bundles in your Modules, but cannot rely on a io field in every Module; for more details see (url here, possibly to this PR)"

And repeat below for Module.

Reply:
Good idea, I more-or-less copied your suggestion.